irohad/ametsuchi/impl/postgres_command_executor.cpp

```diff
@@ -0,0 +1,770 @@
+    }
+
+    /**
+     * Transforms soci statement to CommandResult,
+     * which will have error message generated by error_generator
+     * appended to error received from given result
+     * @param result which can be received by calling execute_
+     * @param error_generator function which must generate error message
+     * to be used as a return error.
+     * Function is passed instead of string to avoid overhead of string
+     * construction in successful case.
+     * @return CommandResult with combined error message
+     * in case of result contains error
+     */
+    template <typename Function>
+    CommandResult makeCommandResult(
+        soci::statement &st,
+        std::string command_name,
+        Function &&error_generator) noexcept {
+      st.define_and_bind();
+      try {
+        st.execute(true);
+      } catch (std::exception &e) {
+        return makeCommandError(error_generator() + "\n" + e.what(),
+                                command_name);
+      }
+      return {};
+    }
+
+    /**
+     * Transforms soci statement to CommandResult,
+     * which will have error message generated exception
+     * Assums that statement query returns 0 in case of success or error code
+     * @param result which can be received by calling execute_
+     * @param error_generator functions which must generate error message
+     * to be used as a return error.
+     * Functions are passed instead of string to avoid overhead of string
+     * construction in successful case.
+     * @return CommandResult with combined error message
+     * in case of result contains error
+     */
+    CommandResult makeCommandResultByReturnedValue(
+        soci::statement &st,
+        std::string command_name,
+        std::vector<std::function<std::string()>> &error_generator) noexcept {
+      uint32_t result;
+      st.exchange(soci::into(result));
+      st.define_and_bind();
+      try {
+        st.execute(true);
+        for (size_t i = 0; i < error_generator.size(); i++) {
+          // Since success code is 0 and error codes starts with 1
+          if (result == i + 1) {
+            return makeCommandError(error_generator[i](), command_name);
+          }
+        }
+        return {};
+      } catch (std::exception &e) {
+        return makeCommandError(e.what(), command_name);
+      }
+    }
+
+    PostgresCommandExecutor::PostgresCommandExecutor(soci::session &sql)
+        : sql_(sql) {}
+
+    void PostgresCommandExecutor::setCreatorAccountId(
+        const shared_model::interface::types::AccountIdType
+            &creator_account_id) {
+      creator_account_id_ = creator_account_id;
+    }
+
+    CommandResult PostgresCommandExecutor::operator()(
+        const shared_model::interface::AddAssetQuantity &command) {
+      auto &account_id = creator_account_id_;
+      auto &asset_id = command.assetId();
+      auto amount = command.amount().toStringRepr();
+      uint32_t precision = command.amount().precision();
```

Comment:
auto?

Reply:
No. Precision is uint8_t, but soci has a bug with uint8_t.